input.go

```diff
@@ -8,7 +8,16 @@ import (
 
 type Input interface {
 	// Text returns the characters to be shaped.
-	Text() []rune
+	//
+	// `itemOffset` is the position of the first character from `text` that will be shaped, and
+	// `itemLength` is the number of character to shape (-1 means the end of the slice).
```

Comment:
My only question is whether using `-1` is really preferable to `len(input)` to specify the end of the list. It doesn't seem significantly more ergonomic to me, but I also haven't written much code that interacts with text shapers. I don't have a strong opinion about this though. I definitely don't mind if it stays this way.

Reply:
Notice that with the convention of the PR, the second `int` is the _length_ of the segment, not the end index. 
Thus, if you want to shape text[2:end_of_the_slice], you would have to pass len(text) - 2 as `itemLength`, which is a bit less ergonomic to just pass `-1`.

We could also flip the convention to provide `(startIndex, endIndex)` if it is more natural...

Reply:
Ah, forgive me. I apparently cannot read. :D I'm fine with -1 as a convention then. It is totally more ergonomic. I would also be fine with the start and end convention, but I see no pressing need for it.

Reply:
Mild preference of start, end as they can be used directly as `for i := start; i < end; i++` loops, without extra `if length == -1` checks

Reply:
Agreed - it seems to be more idiomatic of go as slices etc take that approach